stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -1244,35 +2055,71 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
     ///
     /// - parameter reference: The instance of `NSData` that you wish to wrap. This instance will be copied by `struct Data`.
     public init(referencing reference: __shared NSData) {
+        // This is not marked as inline because _providesConcreteBacking would need to be marked as usable from inline however that is a dynamic lookup in objc contexts.
+        let length = reference.length
+        if length == 0 {
+            _representation = .empty
+        } else {
 #if DEPLOYMENT_RUNTIME_SWIFT
-        let providesConcreteBacking = reference._providesConcreteBacking()
+            let providesConcreteBacking = reference._providesConcreteBacking()
 #else
-        let providesConcreteBacking = (reference as AnyObject)._providesConcreteBacking?() ?? false
+            let providesConcreteBacking = (reference as AnyObject)._providesConcreteBacking?() ?? false
 #endif
-        if providesConcreteBacking {
-            _backing = _DataStorage(immutableReference: reference.copy() as! NSData, offset: 0)
-            _sliceRange = 0..<reference.length
-        } else {
-            _backing = _DataStorage(customReference: reference.copy() as! NSData, offset: 0)
-            _sliceRange = 0..<reference.length
+            if providesConcreteBacking {
+                _representation = _Representation(_DataStorage(immutableReference: reference.copy() as! NSData, offset: 0), count: length)
+            } else {
+                _representation = _Representation(_DataStorage(customReference: reference.copy() as! NSData, offset: 0), count: length)
+            }
         }
         
     }
     
     // slightly faster paths for common sequences
     @inlinable
-    public init<S: Sequence>(_ elements: S) where S.Iterator.Element == UInt8 {
-        let backing = _DataStorage(capacity: Swift.max(elements.underestimatedCount, 1))
-        var (iter, endIndex) = elements._copyContents(initializing: UnsafeMutableBufferPointer(start: backing._bytes?.bindMemory(to: UInt8.self, capacity: backing._capacity), count: backing._capacity))
-        backing._length = endIndex
-        while var element = iter.next() {
-            backing.append(&element, length: 1)
+    @inline(__always)
+    public init<S: Sequence>(_ elements: S) where S.Element == UInt8 {
+        // If the sequence is already contiguous, access the underlying raw memory directly.
+        if let contiguous = elements as? ContiguousBytes {
```

Comment:
Just curious: why is this needed? Are there sequences of `UInt8`s which cannot implement `withContiguousStorageIfAvailable` but can conform to `ContiguousBytes`?

Reply:
Yes: any sequence of `UInt8`'s which does not represent _typed_ memory cannot implement `withContiguousStorageIfAvailable` without handing over a copy, which defeats what we need to do here. `Data` and `UnsafeRawBufferPointer` are both examples of such sequences — both have an element type of `UInt8`, but represent raw memory, and thus cannot implement `withContiguousStorageIfAvailable`.